### PR TITLE
feat(core): support VARCHAR and STRING column type as dedup keys

### PR DESCRIPTION
**Pull request by ideoma**
Allow variable length columns like `VARCHAR`, `STRING` and `BINARY` to be dedup keys:

```SQL
create table x (s varchar, x int, ts timestamp)
timestamp(ts) partition by DAY WAL DEDUP UPSERT KEYS(ts, s);
```

There are a few other minor improvements:

- Support `control iso_timestamp on` in sqllogictests to enable timestamp columns to be conditionally printed as ISO microsecond timestamp
- Support `select wait_wal_table('x');` for testing purposes to wait until all transactions applied to a WAL table

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 191 / 205 (93.17%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64DecodeFunctionFactory.java|23|31|74.19%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/WaitWalTableFunctionFactory.java|21|24|87.50%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|43|46|93.48%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGWireServer.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/DedupColumnCommitAddresses.java|26|26|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionJob.java|56|56|100.00%|
|:large_blue_circle:|io/questdb/cairo/CursorPrinter.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64FunctionFactory.java|4|4|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 192 / 204 (94.12%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64DecodeFunctionFactory.java|23|31|74.19%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/WaitWalTableFunctionFactory.java|21|24|87.50%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGWireServer.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/DedupColumnCommitAddresses.java|26|26|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionJob.java|56|56|100.00%|
|:large_blue_circle:|io/questdb/cairo/CursorPrinter.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|46|46|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64FunctionFactory.java|4|4|100.00%|

**Review by eugenels — COMMENTED** [review]
my review

File: core/src/test/java/io/questdb/test/sqllogictest/AbstractSqllogicTestRunner.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+    @After
+    public void tearDown() throws Exception {
+        LOG.info().$("Finished test ").$(getClass().getSimpleName()).$('#').$(testName.getMethodName()).$();
+        Path p = Path.getThreadLocal(root).concat("db");
+        TestUtils.removeTestPath(root);
+    }
+
+    @Test
+    public void test() {
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/test/java/io/questdb/test/sqllogictest/AbstractSqllogicTestRunner.java:112]: Just to group the files together by directory, more convenient to run

File: core/src/main/java/io/questdb/griffin/engine/functions/bin/Base64DecodeFunctionFactory.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        @Override
+        public Function getArg() {
+            return data;
+        }
+
+        @Override
+        public BinarySequence getBin(Record rec) {
+            CharSequence str = data.getStrA(rec);
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/java/io/questdb/griffin/engine/functions/bin/Base64DecodeFunctionFactory.java:77]: Good point, should be String signature, fixing.

File: core/src/main/java/io/questdb/cutlass/pgwire/PGWireServer.java [hunk 3]
```diff
@@ -154,6 +154,12 @@ public WorkerPool getWorkerPool() {
         return workerPool;
     }
 
+    public void resetQueryCache() {
+        if (typesAndSelectCache != null) {
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/java/io/questdb/cutlass/pgwire/PGWireServer.java:158]: We're hiding "reset" after "close", I think it's ok, "reset" supposed to be used in tests only

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 194 / 205 (94.63%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64DecodeFunctionFactory.java|23|31|74.19%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/WaitWalTableFunctionFactory.java|21|24|87.50%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGWireServer.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/DedupColumnCommitAddresses.java|26|26|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionJob.java|56|56|100.00%|
|:large_blue_circle:|io/questdb/cairo/CursorPrinter.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|46|46|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64FunctionFactory.java|4|4|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 198 / 209 (94.74%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64DecodeFunctionFactory.java|23|31|74.19%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/WaitWalTableFunctionFactory.java|21|24|87.50%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGWireServer.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/DedupColumnCommitAddresses.java|26|26|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionJob.java|59|59|100.00%|
|:large_blue_circle:|io/questdb/cairo/CursorPrinter.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|47|47|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64FunctionFactory.java|4|4|100.00%|

File: core/src/main/c/share/dedup_comparers.h [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    inline int operator()(int64_t col_index, int64_t index_index) const {
+        const T l_val = col_index >= column_top
+                        ? reinterpret_cast<T *>(column_data)[col_index]
+                        : *reinterpret_cast<const T *>(&null_value);
+
+        const auto r_val = reinterpret_cast<T *>(o3_data)[index_index];
+
+        // One of the values can be MIN of the type (null value)
+        // and subtraction can result in type overflow
+        return l_val > r_val ? 1 : (l_val < r_val ? -1 : 0);
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/c/share/dedup_comparers.h:45]: Fuzz tests said no

File: core/src/main/c/share/dedup_comparers.h [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        return l_size - r_size;
+    }
+
+    switch (l_size) {
+        case -1:
+        case 0:
+            return 0;
+
+        default: {
+            return std::memcmp(l_val + sizeof(T), r_val + sizeof(T), l_size * item_size);
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/c/share/dedup_comparers.h:217]: If we don't consider the possibility that lexicographically same UTF-8 string can be represented as different bytes we can use `memcmp`. 
I think it's a fair simplification that would work in practically every case.

File: core/src/test/resources/sqllogictest/test/dedup/varchar_dedup.test [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+# Absolute path
+statement ok
+create table x (s varchar, x int, ts timestamp)
+timestamp(ts) partition by DAY WAL DEDUP UPSERT KEYS(ts, s);
+
+# Insert some data with dups in the single commit
+statement ok
+insert into x(s, x, ts)
+values
+('1234567890', 01, '2020-01-01T00:00:00.000Z'),
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/test/resources/sqllogictest/test/dedup/varchar_dedup.test]: Added non-ascii chars here and in fuzz tests string generation

File: core/src/main/java/io/questdb/cairo/O3PartitionJob.java [hunk 7]
```diff
@@ -809,50 +809,133 @@ private static long getDedupRowsWithAdditionalKeys(
...
+                            dedupCommitAddresses.setReservedValuesSet1(
+                                    addr,
+                                    -1,
+                                    -1,
+                                    -1
+                            );
+                        } else {
+                            // Var len columns
+                            final long oooVarColAddress = oooColumns.get(getPrimaryColumnIndex(i)).addressOf(0);
+                            final long oooVarColSize = oooColumns.get(getPrimaryColumnIndex(i)).size();
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/O3PartitionJob.java:842]: changed to `hi - address`

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 198 / 209 (94.74%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64DecodeFunctionFactory.java|23|31|74.19%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/WaitWalTableFunctionFactory.java|21|24|87.50%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGWireServer.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/DedupColumnCommitAddresses.java|26|26|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionJob.java|59|59|100.00%|
|:large_blue_circle:|io/questdb/cairo/CursorPrinter.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|47|47|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64FunctionFactory.java|4|4|100.00%|

File: core/src/main/c/share/dedup_comparers.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (eugenels) [reply core/src/main/c/share/dedup_comparers.h:45]: hmm, very strange. it should be equivalent. for what types/values does it not work?

[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/c/share/dedup_comparers.h:45]: I added this code:
```C
        int res = l_val > r_val ? 1 : (l_val < r_val ? -1 : 0);
        int res2 = (l_val > r_val) - (l_val < r_val);
        if (res != res2) {
            printf("l_val: %d, r_val: %d, res: %d, res2: %d\n", l_val, r_val, res, res2);
        }
        return res2;
```

and the output on the failing test is:

```
l_val: 151354064, r_val: 151354032, res: 1, res2: 0
```

[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/c/share/dedup_comparers.h]: we found it, it was a bug in long256 comparisons, fixed in [167b7f8](https://github.com/questdb/questdb/pull/4915/commits/167b7f81290cfb9fde4b1e639ca7a73481cff672)

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 199 / 210 (94.76%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64DecodeFunctionFactory.java|23|31|74.19%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/WaitWalTableFunctionFactory.java|21|24|87.50%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGWireServer.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/DedupColumnCommitAddresses.java|26|26|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionJob.java|59|59|100.00%|
|:large_blue_circle:|io/questdb/cairo/CursorPrinter.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|47|47|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64FunctionFactory.java|4|4|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 200 / 211 (94.79%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64DecodeFunctionFactory.java|23|31|74.19%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/WaitWalTableFunctionFactory.java|21|24|87.50%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGWireServer.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/TextMetadataDetector.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/DedupColumnCommitAddresses.java|26|26|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionJob.java|59|59|100.00%|
|:large_blue_circle:|io/questdb/cairo/CursorPrinter.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|47|47|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64FunctionFactory.java|4|4|100.00%|

**Review by javier — APPROVED** [review]
Tested and looks good to me (I did not review code, just functionality)

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 200 / 211 (94.79%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64DecodeFunctionFactory.java|23|31|74.19%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/WaitWalTableFunctionFactory.java|21|24|87.50%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGWireServer.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/TextMetadataDetector.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableUtils.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/DedupColumnCommitAddresses.java|26|26|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3PartitionJob.java|59|59|100.00%|
|:large_blue_circle:|io/questdb/cairo/CursorPrinter.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/FilesFacadeImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|47|47|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bin/Base64FunctionFactory.java|4|4|100.00%|

**Review by eugenels — APPROVED** [review]

**Pull request merged**